hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala

```diff
@@ -148,12 +148,21 @@ private[hudi] object HoodieSparkSqlWriter {
           sparkContext.getConf.registerKryoClasses(
             Array(classOf[org.apache.avro.generic.GenericData],
               classOf[org.apache.avro.Schema]))
-          val schema = AvroConversionUtils.convertStructTypeToAvroSchema(df.schema, structName, nameSpace)
+          var schema = AvroConversionUtils.convertStructTypeToAvroSchema(df.schema, structName, nameSpace)
           sparkContext.getConf.registerAvroSchemas(schema)
           log.info(s"Registered avro schema : ${schema.toString(true)}")
+          var (convertGenRecsToLatestTableSchema, latestSchema) = schemaNeedsConversion(fs, basePath, sparkContext, schema)
 
           // Convert to RDD[HoodieRecord]
-          val genericRecords: RDD[GenericRecord] = HoodieSparkUtils.createRdd(df, schema, structName, nameSpace)
+          var genericRecords: RDD[GenericRecord] = HoodieSparkUtils.createRdd(df, schema, structName, nameSpace)
+          if(convertGenRecsToLatestTableSchema) { // if incoming batch is of old schema but table has evolved schema,
```

Comment:
This should already be taken care by the logic internally right ? If you just set `var schema = convertSchemaIfNeeded(fs, basePath, sparkContext, schema)` and simply return the `latest schema` and pass this schema to `HoodieSparkUtils.createRdd(df, schema, structName, nameSpace)`, internally we should be making sure that the passed schema (which was earlier df.schema) is honored. Can you explain why do we need this "if" condition ?

Reply:
thats what I tried first. but HoodieSparkUtils.createRdd(df with rows serialized with old schema, latest schema (evolved), ...) actually fails during deserializing row (with evolved schema). And so I had to go the longer route of Converting RDD[GenRec] -> Dataset<Row> -> RDD[GenRec]. We can sync up f2f on this to see if we can simplify things. I am all in if it can be simplified.